src/Functional.CQS.AOP.IoC.SimpleInjector.MetricsCapturing.Tests/ConventionBasedDecoratorRegistrationGatewayExtensionsTests.cs

```diff
@@ -0,0 +1,228 @@
+
+		private static void ShouldApplyDecoratorIfConfigurationHasCommandAndQuerySpecificDecorationsEnabled(Container container)
+		{
+			foreach (var type in TestUtility.CQSHandlerContractTypes)
+				container.GetInstance(type).Should().BeOfType(MetricsCapturingTestUtility.MetricsCapturingDecoratorTypeLookupByCQSHandlerContractType[type]);
+		}
+
+		[Theory, SingletonRegistrationWithUniversalDecorationsEnabledArrangement]
+		public void ShouldApplyDecoratorIfConfigurationHasUniversalDecorationsEnabled_SingletonRegistration(Container container) => ShouldApplyDecoratorIfConfigurationHasUniversalDecorationsEnabled(container);
+
+		[Theory, TransientRegistrationWithUniversalDecorationsEnabledArrangement]
+		public void ShouldApplyDecoratorIfConfigurationHasUniversalDecorationsEnabled_TransientRegistration(Container container) => ShouldApplyDecoratorIfConfigurationHasUniversalDecorationsEnabled(container);
+
+		private static void ShouldApplyDecoratorIfConfigurationHasUniversalDecorationsEnabled(Container container)
+		{
+			foreach (var type in TestUtility.CQSHandlerContractTypes)
+				container.GetInstance(type).Should().BeOfType(MetricsCapturingTestUtility.UniversalMetricsCapturingDecoratorTypeLookupByCQSHandlerContractType[type]);
+		}
+
+		[Theory, SingletonRegistrationWithAllDecorationsEnabledArrangement]
+		public void ShouldApplyDecoratorIfConfigurationHasAllDecorationsEnabled_SingletonRegistration(Container container) => ShouldApplyDecoratorIfConfigurationHasAllDecorationsEnabled(container);
+
+		[Theory, TransientRegistrationWithAllDecorationsEnabledArrangement]
+		public void ShouldApplyDecoratorIfConfigurationHasAllDecorationsEnabled_TransientRegistration(Container container) => ShouldApplyDecoratorIfConfigurationHasAllDecorationsEnabled(container);
+
+		private static void ShouldApplyDecoratorIfConfigurationHasAllDecorationsEnabled(Container container)
+		{
+			foreach (var type in TestUtility.CQSHandlerContractTypes)
+				container.GetInstance(type).Should().BeOfType(MetricsCapturingTestUtility.UniversalMetricsCapturingDecoratorTypeLookupByCQSHandlerContractType[type]);
+		}
+
+		[Theory, SingletonRegistrationWithCommandAndQuerySpecificDecorationsEnabledButNoStrategyImplementationsArrangement]
+		public void ShouldNotApplyDecoratorWhenNoStrategyImplementationExists_SingletonRegistration(Container container) => ShouldNotApplyDecoratorWhenNoStrategyImplementationExists(container);
+
+		[Theory, TransientRegistrationWithCommandAndQuerySpecificDecorationsEnabledButNoStrategyImplementationsArrangement]
+		public void ShouldNotApplyDecoratorWhenNoStrategyImplementationExists_TransientRegistration(Container container) => ShouldNotApplyDecoratorWhenNoStrategyImplementationExists(container);
+
+		private static void ShouldNotApplyDecoratorWhenNoStrategyImplementationExists(Container container)
```

Comment:
Should there be test cases for when universal decorations are enabled, but no strategy implementation exists?

I see "but no strategy implementations exist" tests for the other two cases:
- Command/Query-specific decorations are enabled
- All decorations are enabled.

Reply:
Currently, a universal strategy is always defined.  It is either a user-defined type or `NullUniversalMetricsCapturingStrategy` (a [null object](https://en.wikipedia.org/wiki/Null_object_pattern#C#)) depending on the `WithMetricsCapturingDecorator` method overload used.  This is currently done to reduce some code duplication, but this _could_ surprise people when they debug and find out that their handler has been decorated.  Thoughts on possible changes?

Reply:
I'm okay with having "no decoration" be implemented by a decorator that intentionally does nothing. However, I'd recommend documenting that behaviour:
- in the written documentation, so that it's clear to the user that, without implementing their own universal metrics-capturing strategy, the default strategy will do nothing.
- in the unit tests, by demonstrating the scenarios that lead to the null strategy being used.
   - I think adding a test case for "universal decorations are enabled, but no strategy implementation exists" would demonstrate this.

Reply:
After thinking about it a bit, I think that even with notes in the documentation it's not guaranteed that developers will read it.  I'll yield to [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) and _not_ apply the universal decorator if a user-defined implementation is not supplied.